staging/src/k8s.io/client-go/discovery/discovery_client.go

```diff
@@ -137,6 +146,147 @@ type DiscoveryClient struct {
 	restClient restclient.Interface
 
 	LegacyPrefix string
+
+	// Cache groups and resources along with the discovery version.
+	// TODO(seans3): Add lock/mutex here if needed for cache.
+	lock          sync.RWMutex
+	version       string
+	groups        *metav1.APIGroupList
+	resourcesByGV map[string]*metav1.APIResourceList
+}
+
+// Returns true if the discovery client has already retrieved the discovery
+// groups (resources will also be downloaded if discovery is "aggregated"); false otherwise.
+func (d *DiscoveryClient) isCached() bool {
+	return d.version != versionNone
+}
+
+// Returns true of the discovery version negotiated with the
+// api server is "v2beta1" or greater; false otherwise.
+func (d *DiscoveryClient) isAggregated() bool {
+	return d.version == versionV2Beta1
+}
+
+// Returns all the stored discovery groups; empty slice if
+// the groups haven't been cached yet.
+func (d *DiscoveryClient) allGroups() []*metav1.APIGroup {
+	groups := []*metav1.APIGroup{}
+	if d.groups != nil {
+		for _, group := range d.groups.Groups {
+			groups = append(groups, &group)
+		}
+	}
+	return groups
+}
+
+// Returns all the stored discovery resources; empty slice if
+// the resources haven't been cached yet.
+func (d *DiscoveryClient) allResources() []*metav1.APIResourceList {
+	resources := []*metav1.APIResourceList{}
+	if d.resourcesByGV != nil {
+		for _, resourceList := range d.resourcesByGV {
+			resources = append(resources, resourceList)
+		}
+	}
+	return resources
+}
+
+// Returns all the stored "preferred" resources (highest priority
+// version); empty slice if resources haven't been cached yet.
+func (d *DiscoveryClient) preferredResources() []*metav1.APIResourceList {
+	resources := []*metav1.APIResourceList{}
+	if d.groups == nil || d.resourcesByGV == nil {
+		return resources
+	}
+	for _, group := range d.groups.Groups {
+		gv := group.PreferredVersion.GroupVersion
+		if resourceList, found := d.resourcesByGV[gv]; found {
+			resources = append(resources, resourceList)
+		}
+	}
+	return resources
+}
+
+// Returns all the stored "preferred" resources (highest priority
+// version); empty slice if resources haven't been cached yet.
+func (d *DiscoveryClient) preferredNamespacedResources() []*metav1.APIResourceList {
+	filteredResources := []*metav1.APIResourceList{}
+	for _, resourceList := range d.preferredResources() {
+		filteredResourceList := metav1.APIResourceList{}
+		for _, resource := range resourceList.APIResources {
+			if resource.Namespaced {
+				filteredResourceList.APIResources = append(filteredResourceList.APIResources, resource)
+			}
+		}
+		filteredResources = append(filteredResources, &filteredResourceList)
+	}
+	return filteredResources
+}
+
+// Stores the discovery groups. If "aggregated" discovery, resources stored also.
+// TODO(seans3): Need to figure out the mutex strategy here (if needed).
+func (d *DiscoveryClient) cacheDiscovery() error {
```

Comment:
Yes--caching here and elsewhere is confusing. We have to store the discovery/resources in this client because the amount (just groups or everything) received differs based on the returned format.